lib/SILOptimizer/Utils/CFGOptUtils.cpp

```diff
@@ -700,7 +701,73 @@ void swift::completeJointPostDominanceSet(
 
   // Now that we are done, add all remaining must visit blocks to our result
   // list. These are the remaining parts of our joint post-dominance closure.
-  llvm::copy(mustVisitSuccessorBlocks, std::back_inserter(result));
+  llvm::copy(mustVisitSuccessorBlocks, std::back_inserter(completion));
+
+  // Also, compute the postdominance set, which is the union of the user blocks
+  // that are not visited in the above fix point iteration and the
+  // mustVisitSuccessorBlocks.
+  llvm::copy_if(userBlocks, std::back_inserter(postDominators),
+                [&](SILBasicBlock *block) -> bool {
+                  return !visitedBlocks.count(block);
+                });
+  postDominators.append(completion.begin(), completion.end());
+}
+
+void swift::computePostDominatorsOfUsers(
+    SmallPtrSetImpl<SILInstruction *> &users,
+    SmallPtrSetImpl<SILInstruction *> &defs,
+    SmallVectorImpl<SILInstruction *> &postDominators) {
+  // Step 1. Collect the basic blocks of the users and defs.
+  SmallPtrSet<SILBasicBlock *, 4> userBlocks;
+  for (auto *inst : users) {
+    userBlocks.insert(inst->getParent());
+  }
+  SmallPtrSet<SILBasicBlock *, 4> defBlocks;
+  for (auto *inst : defs) {
+    defBlocks.insert(inst->getParent());
+  }
+
+  // Step 2. Compute post-dominators of all user blocks.
+  SmallVector<SILBasicBlock *, 8> completions;
+  SmallVector<SILBasicBlock *, 8> postDominatingBlocks;
+  completeJointPostDominanceSet(userBlocks, defBlocks, completions,
+                                postDominatingBlocks);
+
+  // Step 3. Compute all post-dominating instructions by looking into the
+  // post-dominating blocks.
+  SmallPtrSet<SILBasicBlock *, 8> completionSet(completions.begin(),
+                                                completions.end());
+  for (SILBasicBlock *bb : postDominatingBlocks) {
+    if (completionSet.count(bb)) {
+      // Here, the basic block bb was added to complete the postdominator set
+      // and does not have any users of defs. So, the first instruction of the
```

Comment:
Why can't a completion block have a def?

Reply:
That is an interesting question. I will refer to the reflexive, transitive closure of predecessors of a set of blocks as `pred*(blocks)`.  Essentially, completions are successors of `pred*(userBlocks)` but are not a subset of `pred*(userBlocks)` . If we consider a completion block `c`, `c` is a `succ(pred*(userBlocks))` and `c` is not in `pred*(userBlocks)`. But one thing we know by definition is `defBlocks` are a subset of `pred*(userBlocks)`. This is an implicit precondition (or rather assumption) of this function.  Therefore, `c` cannot be in `defBlocks`. 

However, I think the precondition that  `defBlocks` are a subset of `pred*(userBlocks)` is not really necessary here. And `c` can be a part of `defBlocks` in such cases. Do you think that would cause any problems?

Reply:
Uh OH! If `defBlocks` is always in `userBlocks` then the original post-dom-set utility is making an unexpected assumption that uses never occur before defs in the same block. Given that we allow multiple defBlocks, that's not a safe assumption, it's reasonable for this utility to be called on a phi-web where uses occur before defs.

But you're right, this won't be a problem for your particular use case.